src/lib/appdata.py

```diff
@@ -23,123 +23,67 @@
-        self._stream = stream
-
-    def comment(self, text):
-        self._stream.write("<!--")
-        self._stream.write(text)
-        self._stream.write("-->")
-
-        # Aggravating hack alert!
-        #
-        # appdata files often include a copyright comment between the <?xml
-        # declaration and the root element. In some cases, there are two.
-        # Unfortunately, there is no representation in the SAX API for
-        # whitespace that occurs outside any element, and there's normally
-        # a newline between the comment and the opening element (or the next
-        # comment).
-        #
-        # Work around this by writing a newline in this case; this requires some
-        # coordination between the reader and this writer.
-        if self._reader.outside_root_element:
-            self._stream.write("\n")
-
-    def startCDATA(self, *args):
-        raise NotImplementedError
-
-    def endCDATA(self, *args):
-        raise NotImplementedError
-
-    def endDTD(self, *args):
-        raise NotImplementedError
-
-
-def add_release(in_path_or_stream, out, version, date):
-    reader = AddVersionFilter(version, date, make_parser())
-    lexical_handler = VerbatimLexicalHandler(reader, out)
-    handler = XMLGenerator(out, encoding="UTF-8", short_empty_elements=True)
-    reader.setContentHandler(handler)
-    reader.setProperty(property_lexical_handler, lexical_handler)
-    reader.parse(in_path_or_stream)
-
-
-def add_release_to_file(appdata_path, version, date):
-    with StringIO() as buf:
+import io
+import typing as t
+
+import lxml.etree as ET
+
+
+DEFAULT_INDENT = "  "
+
+
+def _fill_padding(ele: ET.Element):
+    parent = ele.getparent()
+    index = parent.index(ele)
+    level = sum(1 for _ in ele.iterancestors())
+    if len(parent) > 1:
+        if index == len(parent) - 1:
+            if len(parent) > 2:
+                ele.tail = parent[index - 1].tail
+                parent[index - 1].tail = parent[index - 2].tail
+            else:
+                ele.tail = "\n" + DEFAULT_INDENT * (level - 1)
+                parent[index - 1].tail = parent.text
+        else:
+            ele.tail = parent.text
+    else:
+        parent.text = "\n" + DEFAULT_INDENT * level
+        ele.tail = "\n" + DEFAULT_INDENT * (level - 1)
+
+
+def add_release(
+    src: t.Union[t.IO, str],
+    dst: t.Union[t.IO, str],
+    version: str,
+    date: str,
+):
+    tree = ET.parse(src)
+    root = tree.getroot()
+
+    releases = root.find("releases")
+
+    if releases is None:
+        releases = ET.Element("releases")
+        root.append(releases)
+        _fill_padding(releases)
+
+    release = ET.Element("release", version=version, date=date)
+    releases.insert(0, release)
+    _fill_padding(release)
+
+    tree.write(
+        dst,
+        # XXX: lxml uses single quotes for doctype line if generated with
+        # xml_declaration=True,
+        doctype='<?xml version="1.0" encoding="UTF-8"?>',
```

Comment:
Re. the XXX: so is that bad? Is the issue here that the prevailing style for Flatpak apps is to use double-quotes?

Reply:
I usually use `XXX` comments for "we do this weird thing because I don't know a better way to achieve that". Is it misplaced?
>Is the issue here that the prevailing style for Flatpak apps is to use double-quotes?

Yep, almost everyone seems to use double quotes.

Reply:
:shrug: whatever works – I just wasn't sure from the comment whether this was something to fix later, tolerate, or something  else.

Reply:
I'd say "fix when/if becomes possible, tolerate until then".